### DISCUSSION TITLE
postgresql@9.6 is broken

### DISCUSSION THREAD

**Discussion by Splarv** [Homebrew/discussions#3649] [Everyday usage]
When I upgrade brew, I get error:
`==> Upgrading 1 outdated package:
postgresql@9.6 9.6.23 -> 9.6.24_1
Error: postgresql@9.6 has been disabled because it is not supported upstream!`

But, for instance, postgresql@9.5 don't have any error, works smoothly. How to fix postgresql@9.6? I need it for job.

**Comment by SMillerDev** [comment]
There is no fixing, it's disabled on purpose as you can see in the message. Maybe the postgresql app has this version?

**Reply by Splarv** [reply]
I don't see purpose. 9.6.24 still exists on postgresql.org. Yes, it is not supported by postgresql.org, this mean that this is the last version of 9.6.X and there is no upgrade in the feature, but 9.4 and 9.5 too and they are not broken. Why do homebrew artificially prevent upgrading 9.6 to the last version 9.6.24? There is not any reason.

**Reply by SMillerDev** [reply]
> Why do homebrew artificially prevent upgrading 9.6 to the last version 9.6.24? There is not any reason.

The postgresql team determined that the version was now unsupported so homebrew stopped supporting it.

**Reply by Splarv** [reply]
Unssupported 9.6 just mean that there will no more bugfix for this version and 9.6.24 is the last version of 9.6 branch. The last and the most stable version. But you artificially prevent to upgrade to the 9.6.24 and this is only a sign of stupidity. This was not before.

Well, how to really fix this, may be some one need it too:
1. Take ruby file from 9.6.23 version (postgresql@9.6.rb) and change the version and remove "ussoported".
```
class PostgresqlAT96 < Formula
  desc "Object-relational database system"
  homepage "https://www.postgresql.org/"
  url "https://ftp.postgresql.org/pub/source/v9.6.24/postgresql-9.6.24.tar.bz2"
  sha256 "aeb7a196be3ebed1a7476ef565f39722187c108dd47da7489be9c4fcae982ace"
  license "PostgreSQL"

  livecheck do
    url "https://ftp.postgresql.org/pub/source/"
    regex(%r{href=["']?v?(9\.6(?:\.\d+)*)/?["' >]}i)
  end

  keg_only :versioned_formula

  depends_on "openssl@1.1"
  depends_on "readline"

  uses_from_macos "krb5"
  uses_from_macos "libxslt"
  uses_from_macos "openldap"
  uses_from_macos "perl"

  on_linux do
    depends_on "linux-pam"
    depends_on "util-linux"
  end

  def install
    # avoid adding the SDK library directory to the linker search path
    ENV["XML2_CONFIG"] = "xml2-config --exec-prefix=/usr"

    ENV.prepend "LDFLAGS", "-L#{Formula["openssl@1.1"].opt_lib} -L#{Formula["readline"].opt_lib}"
    ENV.prepend "CPPFLAGS", "-I#{Formula["openssl@1.1"].opt_include} -I#{Formula["readline"].opt_include}"

    args = %W[
      --disable-debug
      --prefix=#{prefix}
      --datadir=#{pkgshare}
      --libdir=#{lib}
      --sysconfdir=#{prefix}/etc
      --docdir=#{doc}
      --enable-thread-safety
      --with-bonjour
      --with-gssapi
      --with-ldap
      --with-openssl
      --with-pam
      --with-libxml
      --with-libxslt
      --with-perl
      --with-tcl
      --with-uuid=e2fs
    ]

    # PostgreSQL by default uses xcodebuild internally to determine this,
    # which does not work on CLT-only installs.
    args << "PG_SYSROOT=#{MacOS.sdk_path}" if MacOS.sdk_root_needed?

    system "./configure", *args
    system "make"

    dirs = %W[datadir=#{pkgshare} libdir=#{lib} pkglibdir=#{lib}]

    # Temporarily disable building/installing the documentation.
    # Postgresql seems to "know" the build system has been altered and
    # tries to regenerate the documentation when using `install-world`.
    # This results in the build failing:
    #  `ERROR: `osx' is missing on your system.`
    # Attempting to fix that by adding a dependency on `open-sp` doesn't
    # work and the build errors out on generating the documentation, so
    # for now let's simply omit it so we can package Postgresql for Mojave.
    if DevelopmentTools.clang_build_version >= 1000
      system "make", "all"
      system "make", "-C", "contrib", "install", "all", *dirs
      system "make", "install", "all", *dirs
    else
      system "make", "install-world", *dirs
    end
  end

  def post_install
    (var/"log").mkpath
    postgresql_datadir.mkpath

    # Don't initialize database, it clashes when testing other PostgreSQL versions.
    return if ENV["HOMEBREW_GITHUB_ACTIONS"]

    postgresql_datadir.mkpath
    system "#{bin}/initdb", postgresql_datadir unless pg_version_exists?
  end

  def postgresql_datadir
    var/name
  end

  def postgresql_log_path
    var/"log/#{name}.log"
  end

  def pg_version_exists?
    (postgresql_datadir/"PG_VERSION").exist?
  end

  def caveats
    <<~EOS
      If builds of PostgreSQL 9 are failing and you have version 8.x installed,
      you may need to remove the previous version first. See:
        https://github.com/Homebrew/legacy-homebrew/issues/2510

      This formula has created a default database cluster with:
        initdb #{postgresql_datadir}
      For more details, read:
        https://www.postgresql.org/docs/#{version.major}/app-initdb.html
    EOS
  end

  service do
    run [opt_bin/"postgres", "-D", var/"postgresql@9.6"]
    keep_alive true
    log_path var/"log/postgresql@9.6.log"
    error_log_path var/"log/postgresql@9.6.log"
    working_dir HOMEBREW_PREFIX
  end

  test do
    system "#{bin}/initdb", testpath/"test" unless ENV["HOMEBREW_GITHUB_ACTIONS"]
    assert_equal pkgshare.to_s, shell_output("#{bin}/pg_config --sharedir").chomp
    assert_equal lib.to_s, shell_output("#{bin}/pg_config --libdir").chomp
    assert_equal lib.to_s, shell_output("#{bin}/pg_config --pkglibdir").chomp
  end
end
```
3. Install it by command `brew install ./postgresql@9.6.rb`
4. (optionaly) check that this works by command `brew services start postgresql@9.6`
5. Pin this version to prevent download broken formula from homebrew `brew pin postgresql@9.6`

Well, now I see on every upgrade
```
$ brew upgrade
Error: Not upgrading 1 pinned package:
postgresql@9.6 9.6.24_1
==> No packages to upgrade
```
If you prevent to upgrade to 9.6.24_1, then just remove your file then to clean up my output.

**Comment by SMillerDev** [comment]
> Unssupported 9.6 just mean that there will no more bugfix for this version and 9.6.24 is the last version of 9.6 branch. The last and the most stable version. 

That's not how Wikipedia explains it: https://en.wikipedia.org/wiki/End-of-life_product#Computing. Not to forget that it also lacks any security fixes that were released after this version.

> But you artificially prevent to upgrade to the 9.6.24 and this is only a sign of stupidity. This was not before.

I advise you to read https://github.com/Homebrew/.github/blob/HEAD/CODE_OF_CONDUCT.md#code-of-conduct

**Comment by phil4lif** [comment]
Is 9.5 no longer supported as well?  Home-brew defaults to that version but when I run brew install postgresql@9.5 I get Error: postgresql@9.5 has been disabled because it is not supported upstream!

**Reply by SMillerDev** [reply]
It is not: https://www.postgresql.org/support/versioning/

**Reply by phil4lif** [reply]
Ah I see it only defaulted to 9.5 because that's what I clicked when it suggested after I began to type in the search bar...
